staging/src/k8s.io/kms/encryption/service.go

```diff
@@ -0,0 +1,249 @@
+	baseTransformerFunc func(cipher.Block) value.Transformer
+
+	remoteKMSResponse   *service.EncryptResponse
+	localTransformer    value.Transformer
+	localTransformerErr error
+}
+
+// NewLocalKEKService is being initialized with a key that is encrypted by the
+// remoteService. In the current implementation, the localKEK Service needs to be
+// restarted by the caller after security thresholds are met.
+func NewLocalKEKService(remoteService service.Service) *LocalKEKService {
+	return &LocalKEKService{
+		remoteKMS:           remoteService,
+		baseTransformerFunc: aestransformer.NewGCMTransformer,
+		transformers:        lru.New(cacheSize),
+	}
+}
+
+func (m *LocalKEKService) getTransformerForEncryption(uid string) (value.Transformer, *service.EncryptResponse, error) {
+	// Check if we have a local KEK
+	//  - If exists, use the local KEK for encryption and return
+	//  - Not exists, generate local KEK, encrypt with remote KEK ,store it in cache encrypt the data and return (this can be expensive but only 1 in N calls will incur this additional latency, N being number of times local KEK is reused)
+	m.remoteOnce.Do(func() {
+		m.localTransformerErr = wait.PollImmediateWithContext(context.Background(), time.Second, 5*time.Minute,
+			func(ctx context.Context) (done bool, err error) {
+				key, err := generateKey(32)
+				if err != nil {
+					return false, fmt.Errorf("failed to generate local KEK: %w", err)
+				}
+				block, err := aes.NewCipher(key)
+				if err != nil {
+					return false, fmt.Errorf("failed to create cipher block: %w", err)
+				}
+				transformer := m.baseTransformerFunc(block)
+
+				resp, err := m.remoteKMS.Encrypt(ctx, uid, key)
+				if err != nil {
+					klog.ErrorS(err, "failed to encrypt local KEK with remote KMS", "uid", uid)
+					return false, nil
+				}
+				if err = validateRemoteKMSResponse(resp); err != nil {
+					return false, fmt.Errorf("response annotations failed validation: %w", err)
+				}
+				m.remoteKMSResponse = copyResponseAndAddLocalKEKAnnotation(resp)
+				m.localTransformer = transformer
+				m.transformers.Add(resp.Ciphertext, transformer)
+				return true, nil
+			})
+	})
+	return m.localTransformer, m.remoteKMSResponse, m.localTransformerErr
+}
+
+func copyResponseAndAddLocalKEKAnnotation(resp *service.EncryptResponse) *service.EncryptResponse {
+	annotations := make(map[string][]byte, len(resp.Annotations)+numAnnotations)
+	for s, bytes := range resp.Annotations {
+		s := s
+		bytes := bytes
+		annotations[s] = bytes
+	}
+	annotations[referenceKEKAnnotationKey] = resp.Ciphertext
+
+	return &service.EncryptResponse{
+		KeyID:       resp.KeyID,
+		Annotations: annotations,
+	}
+}
+
+// Encrypt encrypts the plaintext with the localKEK.
+func (m *LocalKEKService) Encrypt(ctx context.Context, uid string, pt []byte) (*service.EncryptResponse, error) {
+	transformer, resp, err := m.getTransformerForEncryption(uid)
+	if err != nil {
+		klog.V(2).InfoS("encrypt plaintext", "id", uid, "err", err)
```

Comment:
Can we change all `id` to `uid` to match the gRPC API?

```suggestion
		klog.V(2).InfoS("encrypt plaintext", "uid", uid, "err", err)
```